AsyncNats/Messages/NatsMessageParser.cs

```diff
@@ -2,80 +2,407 @@
             var messageCount = 0;
             var reader = new SequenceReader<byte>(buffer);
             while (messageCount<outMessages.Length)
             {
                 var previousPosition = reader.Consumed;
                 if (!reader.TryReadTo(out ReadOnlySpan<byte> line, (byte) '\n')) break;
                 line = line.Slice(0, line.Length - 1); // Slice out \r as well (not just \n)
+                
+                INatsServerMessage? message = null;
+                bool isInlined = false;
+                switch (line[0])
+                {
+                    case (byte)'M': message = ParseMessage(line, ref reader, out isInlined); break;
+                    case (byte)'H': message = ParseMessageWithHeader(line, ref reader,out isInlined); break;
+                    case (byte)'+': message = ParseOk(); break;                    
+                    case (byte)'I': message = ParseInformation(line); break;
+                    case (byte)'-': message = ParseError(line); break;
+                    case (byte)'P': message = line[1] == (byte)'I' ? ParsePing():ParsePong(); break;
+                    default:
+                        throw new ProtocolViolationException($"Unknown message {Encoding.UTF8.GetString(line)}");
+                }
 
-                ParseMessage? parseMessage = null;
-                // Not a foreach as it generates an enumerator which allocates bytes
-                // ReSharper disable once ForCanBeConvertedToForeach
-                for (var i = 0; i < _messageParsers.Length; i++)
+                if (!isInlined)
                 {
-                    if (line.StartsWith(_messageParsers[i].Item1))
+                    if (message == null)
                     {
-                        parseMessage = _messageParsers[i].Item2;
+                        // Not enough information to parse the message
+                        reader.Rewind(reader.Consumed - previousPosition);
                         break;
                     }
+
+                    outMessages[messageCount] = message;
+                    messageCount++;
                 }
-                
-                if (parseMessage == null) throw new ProtocolViolationException($"Unknown message {Encoding.UTF8.GetString(line)}");
-                var message = parseMessage(_memoryPool, line, ref reader);
-                if (message == null)
+                else
                 {
-                    // Not enough information to parse the message
-                    reader.Rewind(reader.Consumed - previousPosition);
-                    break;
+                    inlined++;
                 }
-
-                outMessages[messageCount] = message;
-                messageCount++;
                
             }
 
             consumed = reader.Consumed;
             return messageCount;
         }
+
+        public NatsMsg? ParseMessage(in ReadOnlySpan<byte> line, ref SequenceReader<byte> reader, out bool inlined)
```

Comment:
> I might look at it again at a later point [have a headache], but overall looks good. Minor points here and there.
> 
> It might be an idea to somehow clone the InlineSubscriptions from a concurrentdictionary to a normal dictionary (and use an overwrite strategy) after a change. Concurrent dictionaries are much heavier to read compared to normal dictionaries (that said I suspect even a "hasInlineSubs" check would be nice for a normal dictionary).

That's ok, let me take a run on your comments and suggestions.

Reply:
I'm not a fan too. The reason here was to avoid the indirection of calls made via interface methods, and potentially allow inlining, since the message parser method ( the most used and critical ) now access the reference dictionary of inline subscriptions.
Let me know if I should explore alternatives, I'm leaving as is for now.